op-challenger/game/keccak/matrix/matrix.go

```diff
@@ -17,10 +18,61 @@ type StateMatrix struct {
 }
 
 var (
-	ErrInvalidMaxLen = errors.New("invalid max length to absorb")
-	uint256Size      = 32
+	ErrInvalidMaxLen            = errors.New("invalid max length to absorb")
+	ErrIncorrectCommitmentCount = errors.New("incorrect number of commitments for input length")
+	uint256Size                 = 32
 )
 
+// VerifyPreimage checks the claimed state commitments are correct for the provided preimage data.
+// Returns nil if the preimage is valid, otherwise returns a [types.Challenge] proving the fault.
+func VerifyPreimage(data io.Reader, commitments []common.Hash) (*types.Challenge, error) {
+	s := NewStateMatrix()
+	m := s.PackState()
+	var prestate types.Leaf
+	for i := 0; ; i++ {
+		unpaddedLeaf, err := s.absorbNextLeafInput(data)
+		if err != nil && !errors.Is(err, io.EOF) {
+			return nil, fmt.Errorf("failed to verify inputs: %w", err)
+		}
+		isEOF := errors.Is(err, io.EOF)
```

Comment:
personal legibility preference, feel free to ignore
```suggestion
		isEOF := errors.Is(err, io.EOF)
		if err != nil && !isEOF {
			return nil, fmt.Errorf("failed to verify inputs: %w", err)
		}
```